libvmi/driver/xen/altp2m.c

```diff
@@ -24,6 +24,37 @@
 #include "driver/xen/xen.h"
 #include "driver/xen/xen_private.h"
 
+status_t xen_altp2m_init (vmi_instance_t vmi, uint64_t *init_memsize)
+{
+    int rc;
+    xen_instance_t *xen = xen_get_instance(vmi);
+    xc_interface * xch = xen_get_xchandle(vmi);
+    domid_t domain_id = xen_get_domainid(vmi);
+
+    xc_dominfo_t info = { 0 };
+    if ( 1 == xen->libxcw.xc_domain_getinfo(xch, domain_id, 1, &info) && info.domid == domain_id)
+        *init_memsize = info.max_memkb;
+    else
+        *init_memsize = 0;
+
+    rc = xen->libxcw.xc_domain_setmaxmem(xch, domain_id, ~0);
+    if (rc < 0)
+        return VMI_FAILURE;
+
+    return VMI_SUCCESS;
+}
+
+status_t xen_altp2m_deinit (vmi_instance_t vmi, uint64_t init_memsize)
+{
+    xen_instance_t *xen = xen_get_instance(vmi);
+    xc_interface * xch = xen_get_xchandle(vmi);
+    domid_t domain_id = xen_get_domainid(vmi);
+
+    xen->libxcw.xc_domain_setmaxmem(xch, domain_id, init_memsize);
```

Comment:
This is a hack copied from DRAKVUF that I would not like to repeat in LibVMI. The only way to reverse the domain to its initial state would be get the return value of `XENMEM_maximum_reservation` hypercall before it got changed, and it is not currently exposed through `libxc`.

Reply:
Never mind, some more grepping shows that `xc_domain_getinfo` returns that same initial value `XENMEM_maximum_reservation` does. I guess `XENMEM_maximum_reservation` is just kept around for compatibility reasons in case some toolstack still uses it.